common/src/test/java/org/conscrypt/javax/net/ssl/SSLEngineVersionCompatibilityTest.java

```diff
@@ -809,6 +807,127 @@ public boolean matches(SNIServerName sniServerName) {
+            sourceBuffers = addInvalidBuffers(sourceBuffers, offset, bufferSize);
+        }
+
+        // Ensure there is no pending outbound data or encrypted data and handshaking is complete.
+        ByteBuffer tlsBuffer = ByteBuffer.allocateDirect(src.getSession().getPacketBufferSize());
+        SSLEngineResult result = src.wrap(EMPTY_BUFFER, tlsBuffer);
+        assertEquals(Status.OK, result.getStatus());
+        assertEquals(HandshakeStatus.NOT_HANDSHAKING, result.getHandshakeStatus());
+        assertEquals(0, result.bytesConsumed());
+        assertEquals(0, result.bytesProduced());
+
+        // Ensure there is no pending inbound data
+        result = dst.unwrap(EMPTY_BUFFER, tlsBuffer);
+        assertEquals(Status.BUFFER_UNDERFLOW, result.getStatus());
+        assertEquals(0, result.bytesConsumed());
+        assertEquals(0, result.bytesProduced());
+
+        // Send the data.  wrap() should consume as many source buffers as needed but
+        // never generate more than one full packet buffer of TLS data, and so unwrap()
+        // should only be needed once per loop iteration.
+        int consumed = 0, produced = 0;
+        ByteBuffer destBuffer = ByteBuffer.allocate(dataSize);
+        while (consumed < dataSize) {
+            String message = String.format("sendData: dataSize=%d, bufSize=%d, offset=%d",
+                    dataSize, bufferSize, offset);
+
+            tlsBuffer.clear();
+            result = src.wrap(sourceBuffers, offset, length, tlsBuffer);
+            assertEquals(message, Status.OK, result.getStatus());
+            consumed += result.bytesConsumed();
+
+            tlsBuffer.flip();
+            result = dst.unwrap(tlsBuffer, destBuffer);
+            assertEquals(message, Status.OK, result.getStatus());
+            produced += result.bytesProduced();
+        }
+        assertEquals(dataSize, consumed);
+        assertEquals(dataSize, produced);
+
+        // Compare source and destination data.
+        destBuffer.flip();
+        // destBuffer is non-direct so will always have a backing array
+        assertArrayEquals(sourceData, destBuffer.array());
+    }
+
+    /**
+     * Tests the multiple {@link ByteBuffer} cases of {@code wrap())} by sending blocks of
+     * application data split into arrays of ByteBuffers of different sizes.
+     *
+     * The main intention is to check that regardless of how the data is structured in
+     * buffers, each call to wrap() always generates a single TLS record that is smaller
+     * than the maximum allowed size, see https://github.com/google/conscrypt/issues/929
+     */
+    @Test
+    public void multipleBuffersOfDifferentSizes() throws Exception {
+        TestSSLEnginePair pair = TestSSLEnginePair.create();
+        SSLSession session = pair.client.getSession();
+        int appBufSize = session.getApplicationBufferSize();
+
+        int[] dataSizes = new int[] { 12, 512, 555, 1500, 8192, appBufSize, 5 * appBufSize};
```

Comment:
There are several comparisons against SSL3_RT_MAX_PLAIN_LENGTH. Should it(and close to it values) also appear here?

Reply:
The implementation code uses `SSL3_RT_MAX_PLAIN_LENGTH`, but that's exposed to client code via [`getApplicationBufferSize()`](https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ActiveSession.java#L280), hence the `appBufSize` entries here and in `bufferSizes`.